packages/avocode-email-tagsinput/test/e2e/basic.spec.js

```diff
@@ -0,0 +1,116 @@
+import puppeteer from 'puppeteer'
+import { expect } from 'chai'
+import {
+  URL,
+  tagSubmitKeys,
+  tagsInputFieldSelector,
+  getAnchorAndFocus,
+  addEmailTag,
+  assertEmailTagCount,
+  assertInputText,
+} from './test-utils'
+
+describe('Basic Email Tags Input', () => {
+  let browser
+  let page
+
+  beforeAll(async () => {
+    browser = await puppeteer.launch(browserOptions)
+  })
+
+  beforeEach(async () => {
+    page = await browser.newPage()
+    await page.goto(URL)
+  })
+
+  afterEach(async () => {
+    await page.close()
+  })
+
+  afterAll(async () => {
+    await browser.close()
+  })
+
+  it('should render editor', async () => {
+    const editor = await page.waitForSelector(tagsInputFieldSelector)
+    expect(editor).to.exist
+  })
+
+  describe('user typing', () => {
+    it('should insert text', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await assertInputText(page, 'abcd')
+    })
+
+    it('should delete text with backspace', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('Backspace')
+      await assertInputText(page, 'abc')
+    })
+
+    it('should delete text with delete', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('ArrowLeft')
+      await page.keyboard.press('Delete')
+      await assertInputText(page, 'abc')
+    })
+
+    it('should navigate one char left', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('ArrowLeft')
+      const [ anchorOffset ] = await getAnchorAndFocus(page)
+      expect(anchorOffset).to.equal(3)
+    })
+  })
+
+  describe('user typing different email edge cases', () => {
```

Comment:
I think this whole `describe` block should be deleted. You can just have two specs testing valid email and invalid email and that's it.

The logic behind validating email address is implemented in `isValueValidEmail` and you should probably write unit tests for utils. It's a different concern.

Reply:
What I mean is that there should be unit test testing file `lib/utils/index` where you would write unit tests for testing input / output of these utils.

In E2E tests I think it's OK to test for valid / invalid cases. (We're not writing [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_%28software%29) here so I think we're safe).